server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -0,0 +1,156 @@
+        super(delayMs);
+        this.lock = lock;
+    }
+
+    /*
+     * Force completing the delayed operation, if not already completed.
+     * This function can be triggered when
+     *
+     * 1. The operation has been verified to be completable inside tryComplete()
+     * 2. The operation has expired and hence needs to be completed right now
+     *
+     * Return true iff the operation is completed by the caller: note that
+     * concurrent threads can try to complete the same operation, but only
+     * the first thread will succeed in completing the operation and return
+     * true, others will still return false
+     */
+    public boolean forceComplete() {
+        if (completed.compareAndSet(false, true)) {
+            // cancel the timeout timer
+            cancel();
+            onComplete();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Check if the delayed operation is already completed
+     */
+    public boolean isCompleted() {
+        return completed.get();
+    }
+
+    /**
+     * Call-back to execute when a delayed operation gets expired and hence forced to complete.
+     */
+    public abstract void onExpiration();
+
+    /**
+     * Process for completing an operation; This function needs to be defined
+     * in subclasses and will be called exactly once in forceComplete()
+     */
+    public abstract void onComplete();
+
+    /**
+     * Try to complete the delayed operation by first checking if the operation
+     * can be completed by now. If yes execute the completion logic by calling
+     * forceComplete() and return true iff forceComplete returns true; otherwise return false
+     * <br/>
+     * This function needs to be defined in subclasses
+     */
+    public abstract boolean tryComplete();
+
+    /**
+     * Thread-safe variant of tryComplete() and call extra function if first tryComplete returns false
+     * @param action else function to be executed after first tryComplete returns false
+     * @return result of tryComplete
+     */
+    boolean safeTryCompleteOrElse(Action action) {
+        lock.lock();
+        try {
+            if (tryComplete()) return true;
+            else {
+                action.run();
+                // last completion check
+                return tryComplete();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Thread-safe variant of tryComplete()
+     */
+    boolean safeTryComplete() {
+        lock.lock();
+        try {
+            if (isCompleted()) return false;
+            else return tryComplete();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * run() method defines a task that is executed on timeout
+     */
+    @Override
+    public void run() {
+        if (forceComplete())
+            onExpiration();
+    }
+
+    @FunctionalInterface
+    public interface Action {
+        void run();
```

Comment:
nit: Should we call the method as `apply`, more synonymous to Java Function library?
 
```suggestion
        void apply();
```